pkg/controller/operators/catalog/operator.go

```diff
@@ -1857,11 +1859,19 @@ func (o *Operator) ExecutePlan(plan *v1alpha1.InstallPlan) error {
 					return errorwrap.Wrapf(err, "error decoding %s object to an unstructured object", step.Resource.Name)
 				}
 
-				// Get the resource from the GVK.
+				// Get the GVK from the resource.
 				gvk := unstructuredObject.GroupVersionKind()
+				// Then check if the API for this dynamic resource is available on-cluster, before creating the resource, by querying discovery.
+				// This ensures CRs created via the dynamic client are created after the corresponding CRD is accepted and established on the cluster to avoid potential race conditions.
+				// (Default) 2 minute timeout for the required API to appear in discovery before failing the installplan
+				if timeout(plan.CreationTimestamp, o.dynamicResourceTimeout) {
```

Comment:
Should the timeout begin when we first attempt to create the CR (or CRD), rather than the plan itself?

Reply:
That probably is more accurate. I'll look into using a timestamp on a per-installplan basis

Reply:
Per installplan step

Reply:
This is a little tricky to implement though. Simply using a local timestamp in `ExecutePlan()` wouldn't work because it would get GC'd everytime the function exits, and so the timeout would never occur. This timestamp needs to persist somehow between reconciliation cycles. I see two possible options:
* use a `map[*v1alpha1.InstallPlan]time.Time` at the package level to set/lookup timestamps per installplan. This would add package-level state though and the map would grow linearly in the number of processed installplans. Even though I implemented this locally I don't think its a good solution. 
* add an optional field to the InstallPlan, or specifically the `InstallPlan.Status.Step` resource called `Timestamp` that we can set/look at from this portion of the code. 

Does the second option sound correct? Are there any other options I'm missing @njhale?